### PR TITLE
ENH: add buffer records overflow metrics

### LINKED ISSUE #2167: Add buffer overflow metric in AbstractBuffer
**Is your feature request related to a problem? Please describe.**
Our AbstractBuffer is missing a metric that reflect the total number of records/events that have been abandoned due to buffer size full or write timeout.

**Describe the solution you'd like**
Add `recordsOverflow` counter that tracks the total number of records failed to be written into the buffer.

**Describe alternatives you've considered (Optional)**
N/A

**Additional context**

### PR DESCRIPTION
**Pull request by chenqi0805**
### Description
Added recordsOverflow metric as common metric in AbstractBuffer
 
### Issues Resolved
Resolves #2167 
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/MetricNames.java [hunk 1]
```diff
@@ -57,6 +57,11 @@ private MetricNames() {}
      */
     public static final String RECORDS_PROCESSED = "recordsProcessed";
 
+    /**
+     * Metric representing the number of records failed to be written into the buffer.
+     */
+    public static final String RECORDS_OVERFLOW = "recordsOverflow";
```
Comment (dlvenable) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/metrics/MetricNames.java:63]: Perhaps `recordsWriteFailed` would be clearer? We are not overflowing the buffer, but instead failed to write to the buffer.

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/buffer/AbstractBufferTest.java [hunk 2]
```diff
@@ -198,6 +212,22 @@ public void testWriteAllTimeoutMetric() throws TimeoutException {
         Assert.assertEquals(1.0, timeoutMeasurements.get(0).getValue(), 0);
     }
 
+    @Test
+    public void testWriteAllRecordsOverflowMetric() {
+        // Given
+        final AbstractBuffer<Record<String>> abstractBuffer = new AbstractBufferRuntimeExceptionImpl(testPluginSetting);
```
Comment (dlvenable) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/buffer/AbstractBufferTest.java:218]: You could accomplish this via mocking:

```
final AbstractBuffer<Record<String>> abstractBuffer = mock(AbstractBuffer.class);
when(abstractBuffer).doWriteAll(any(Collection.class), anyInteger()).thenThrow(RuntimeException.class);
```

Then you can remove the `AbstractBufferRuntimeExceptionImpl` class.

Comment (dlvenable) [reply data-prepper-api/src/test/java/org/opensearch/dataprepper/model/buffer/AbstractBufferTest.java:218]: I see what you mean. For future reference, you can configure Mockito to call real methods.

```
when(abstractBuffer.writeAll(any(Collection.class), anyInteger()).thenCallRealMethod();
```

**Review by dlvenable — APPROVED** [review]

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/buffer/AbstractBufferTest.java [hunk 3]
```diff
@@ -198,6 +212,22 @@ public void testWriteAllTimeoutMetric() throws TimeoutException {
...
+        final Collection<Record<String>> testRecords = Arrays.asList(
+                new Record<>(UUID.randomUUID().toString()), new Record<>(UUID.randomUUID().toString()));
+
+        // When/Then
+        Assert.assertThrows(RuntimeException.class, () -> abstractBuffer.writeAll(testRecords, 1000));
+
+        final List<Measurement> recordsWriteFailedMeasurements = MetricsTestUtil.getMeasurementList(
+                new StringJoiner(MetricNames.DELIMITER).add(PIPELINE_NAME).add(BUFFER_NAME).add(MetricNames.RECORDS_WRITE_FAILED).toString());
+        Assert.assertEquals(1, recordsWriteFailedMeasurements.size());
+        Assert.assertEquals(2.0, recordsWriteFailedMeasurements.get(0).getValue(), 0);
```
Comment (asifsmohammed) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/buffer/AbstractBufferTest.java:228]: QQ. What's the thrid parameter `0` here in assertEquals?

**Review by asifsmohammed — APPROVED** [review]

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/buffer/AbstractBufferTest.java (hunk 3 above)
Comment (dlvenable) [reply data-prepper-api/src/test/java/org/opensearch/dataprepper/model/buffer/AbstractBufferTest.java:228]: I do find Hamcrest helpful for readability in such scenarios.

```
assertThat(recordsWriteFailedMeasurements.get(0).getValue(), closeTo(2.0, 0));
```

**Pull request merged**